Standardize header metadata.

DinoBektesevic opened the pull request:
This is a first attempt at taking the various varied FITS headers and standardizing a select number of header keywords that we can use in our header metadata DB table.

The header keywords we standardize on here will set our table schema and our query tool interface but there are some outstanding issues with standardizing the WCSs. 
If we can not easily separate WCS data into a standardized set of columns querying that table will be much harder but it should still be possible to save a subset of keywords (center pixel values and coordinates) and the whole WCS as a pickled blob. 

This would not be optimal, as optimal as storing all of these values separately,  I suspect. Ergo, this draft PR to give some insight into what the difficulties are.

DinoBektesevic committed:
Standardize header metadata.

Attempt to use astro_metadata_translator and Astropy WCS to
standardize the FITS metadata we want to push to our DB.

DinoBektesevic committed:
Save our own reference pixel values instead.

Separate image center pixel, image corner pixel and
radius (between the two) into separate values and store those.

TODO: store the actual WCS as a blob data and move on to
astro metadata translator.

DinoBektesevic committed:
Ingest more metadata from header.

Fix the Django metadata model.

DinoBektesevic committed:
OOP processing scripts.

Prettify code. Allow for future expansion of processing, potentially,
compressed, fits archives.

Create small and large thumbnails for gallery.

Add some extra practical functionality to TemporaryUploadedFile.

archives

DinoBektesevic committed:
Split code into files and add some documentation (and notes).

mrawls commented on trail/upload/models.py:
```diff
@@ -3,29 +3,29 @@
+
+    obs_lon = models.FloatField("observatory longitude (deg)", null=False)
+    obs_lat = models.FloatField("observatory latitude (deg)", null=False)
+    obs_height = models.FloatField("observatory ASL height (m)", null=False)
```
Not sure if this is legitimately "above sea level," let's just call it "observatory height (m)"

mrawls commented on trail/upload/models.py:
```diff
@@ -3,29 +3,29 @@
+    datetime_begin = models.DateTimeField("UTC at exposure start.", null=False)
+    datetime_end = models.DateTimeField("UTC at exposure end.", null=False)
+
+    radius = models.FloatField("distance between center and corner pixel", null=False)
```
prefix this and all the following things with wcs, e.g., `wcs_radius`

mrawls commented on trail/upload/process_uploads/processors.py:
```diff
@@ -0,0 +1,286 @@
+            # translator names are easy everyday names
+            # the rest I don't know how they'll vary over different
+            # instruments but I can assume potentially by a lot?
+            standardizedKeys["name"] = translated.name
```
Call this `metadata_translator_name` or something a bit more specific so people don't misconstrue it as, e.g., the name of a satellite or other target in the image

mrawls commented:
Metadata still missing: band or filter, exposure duration (grab independently instead of doing end - start if feasible), some kind of processed/reduced flag (probably either "yes" or "unknown")

DinoBektesevic committed:
Remove accidentally commited file.

DinoBektesevic committed:
Review fixes, add tests, fix bugs.

Fixes for PR review comments.
Add a minimal test dataset.
Add tests.
Fix for naive timezone DB error.
Fix for table migration error.
Remove unneccessary (I think) migration directories from apps that
don't do any database interaction.

DinoBektesevic commented:
> Metadata still missing: band or filter, exposure duration (grab independently instead of doing end - start if feasible), some kind of processed/reduced flag (probably either "yes" or "unknown")

Added filter and exposure time for `astro_metadata_translator` recognized instruments. 

Added test dataset to get an overview of where functionality is right now. Coverage right now is about 50/50 of the dataset you compiled. Images in the dataset have been cropped for size but the WCS values have been shifted to maintain positional accuracy (approximately). 

I started gnawing at getting more instruments supported today but nothing that I would like to commit quite yet. I think our discussion was on-point and that whether we want or no we will have to create a `HeaderStandardizer` and a `ImageStandardizer` classes that do the atomic work on a single header and image and then see if `FitsProcessor` can be made to support both single and multi-extension fits files. If not a bit of a redesign of the `FitsProcessor` into a `SingleExtensionFitsProcessor` and a `MultiExtensionFitsProcessor` might be needed so I don't want to promote this into a full PR quite yet.

DinoBektesevic committed:
Support more types of FITS files.

Add Processors - classes that understand the layout of differet
    FITS files and how they should be processed.
Add Standardizers- classes that understand the different types of
    HDUs and are extract our database model keys from them.
Add dataclasses - wider adoption unclear but they do make life
    significantly easier when it comes to comparing and validating
    various collections of standardized keys.
Fix the rest of the upload app to interface the new functionality.
Add tests.
Add docstring.

DinoBektesevic commented:
Ok, this is now getting much better. I think someone may take a look at what is in this code and not feel offended. 

Added header standardizer classes, essentially maps between what we want in our DB and what we can find in particular instrument's header

Processors are essentially recipes of how particular file should be processed. For example, a compressed archive needs to be uncompressed, "unarchived" then each file needs to be processed; but is this particular file Rubin calexp - then the image is only in 1st HDU - or is it a DECam NOAO image - then all HDUs have headers and data - is it an SDSS.... 



To do:
* rework the DB schema to be able to accept focal planes.        
    * Different pipelines produce different types of FITS files. Some store all CCDs from a focal plane in the same FITS, some split them up. The fake DB model right now accepts everything as a single CCD. We either unravel multiext fits into individual CCDs on a generic basis or we implement a better schema. 
    * DB schema could also use a lot of work in terms of recording uploading times, ips, etc...
*  Fix the `astro_metadata_translator` issues with `Header` https://github.com/lsst/astro_metadata_translator/pull/53
* Fix science validity of the parsed headers.     
    * As-is we are not parsing MOA-II fits file WCS data correctly and it's unclear how to. Perhaps this is a good test case for Astrometry.NET integration. Other headers insert data into the DB, but timezones and other details could be an issue. 
    * Fix timezone-naive `Time.iso` output.     
Format for the time-zone aware `datetime` objects seem to be a bit different and Astropy does not seem to return timezone encoded strings.
    * Standardize database schema values, as is instruments, telescopes, filters etc all kind of return something different. For certain things we can guarantee consistency (f.e. what is in telescope, what in instrument etc. for others, i.e. filter or science-program, we can do our best to be as close to something standard but really can not be consistent). Perhaps make these fields something like "identifier" and then adopt the instruments conventional naming scheme (f.e. `run-calcol-filter-field` for SDSS or `run-field-filter-chip` for MOA-II etc.)
    * This is literally an easy free-for-all problem, it just takes emailing people in cases like MOA-II or, for bigger instruments, reading documentation. Las Cumbres standardizer even has a comment to paper with the lookup table.
* Figure out what I want to do with the dataclasses
    * they are very helpful when debugging etc and it is nice when you don't have to traverse highly nested dicts, but their function would ideally be handled by the model classes. Unfortunately the schema is not very reflective of processing steps and will complain for mandatory values for example.
    * no tests for now, until we decide we like or don't like them
* Add a `register` method to processors and standardizers? 
    * gets rid of the import funkery that has to happen right now.
* Add a (manual, or automatic) priority levels to processors/standardizers
    * or remember procs./stand. that volunteered and try to reprocess on fail....
* better naming than `process_upload.upload_processor.process` etc...
    * help appreciated, I am not imaginative
* Consider getting rid of `from*` and `get*` methods on these classes and make instantiation do that bit of work too (makes it prettier?)
* Perhaps a mixin class to standardize all these similar method names and reduce amount of code duplication.
* flakify
* flesh out the tests
    *  as is we essentially just run the entire pipeline end-to-end so we excercise the code, but we should really make it more atomic
* fix how gallery works before this is merged.

DinoBektesevic committed:
Rename modules and cleanup.

DinoBektesevic committed:
Add full focal plane thumbnails for DECam Community Pipelines products.

DinoBektesevic committed:
Add docstrings, rename methods to be easier to understand.

DinoBektesevic committed:
Add prioritization to standardizers and processors.

DinoBektesevic committed:
Add tests for correct matching with priorities.

Fix the ugly test data yaml file.
Replace the DECam test data with newer one that will more accurately
represent DECam data.
Fix a slight dataclass issue once we changed the format of the
test dataset yaml file.

DinoBektesevic committed:
Reorganization and minor fixes.

Move SingleExtensionFits and MultiExtensionFits to processors.
Move primary header stuff to FitsProcessor as common pattern.
Fix minor problems with priority values.
Move FitsProcessor to UploadProcessor level.

DinoBektesevic committed:
Move .standardizeHeader to FitsProcessor.
Get rid of the long and short output format.

Simplify dataclasses.
Fix tests, fix comment.

mrawls commented on trail/upload/process_uploads/processors/decam_processor.py:
```diff
@@ -0,0 +1,360 @@
+    12: {"indices": ((2, 11), (3, 11), (4, 11)),
+         "names":   ( "N31",   "N30",   "N29"), "rtype": "even"},
+}
+"""A row-based layour of the DECam focal plane science detectors."""
```
Double check this against the actual layout at, e.g., http://www.ctio.noao.edu/noao/node/2250. The preview you showed today looked like some things weren't in the right spot, but it's a very nice plot thumbnail for a multi-extension-FITS file.

mrawls replied:
Did you have a chance to double check this now puts the CCDs in the correct spot?

mrawls commented on trail/upload/models.py:
```diff
@@ -3,29 +3,31 @@
+    datetime_begin = models.DateTimeField("UTC at exposure start.", null=False)
+    datetime_end = models.DateTimeField("UTC at exposure end.", null=False)
+    exposure_duration = models.FloatField("exposure time (s)")
+    physical_filter = models.CharField("physical filter", max_length=30)
```
Consider renaming to `filter_name` since this isn't specifically the physical piece of glass a la Rubin Science Pipelines.

mrawls replied:
Still prefer `filter_name` if you're OK with that

DinoBektesevic committed:
Method storeHeader is not an abstract method.

DinoBektesevic committed:
Flakeify

DinoBektesevic committed:
Fix order of CCD identifiers in DECAMFocalPlane processor.

DinoBektesevic committed:
Update DB schema to allow storing multi-extension FITS files.

Storing multi-ext FITS files is complicated because it's
not trivial to unravel them (leaving potentially many images
without trails) and it's not easy to store only the relevant
data. As was discussed, we store the whole file, and live
with a little bit of non-user-friendliness.

Fix the poor naming scheme of UploadProcessor attributes,
what was `upload` is `fileWrapper`, what is `upload` now
is the UploadInfo model.

Add UploadInfo model, which can be used to track what IP
uploaded which data and when.

Reset migration history.

Refactor the view and intiial processing code, punting
much of the functionality into the UploadProcessor.

Replace metadata_translator_name column by standardizer
and processor name columns. Add special naming convention
for astro_metadata_translator classes. Default the rest
to their class names.

Fix tests.

mrawls commented on trail/upload/models.py:
```diff
@@ -3,29 +3,49 @@
+
+    # verbose_names should be lowercase, Django will capitalize
+    # https://docs.djangoproject.com/en/3.1/topics/db/models/#verbose-field-names
+    processor_name = models.CharField("name of used translator.", max_length=20, null=False)
```
Please standardize the verbose names without periods, unless you think they enhance clarity. This one in particular could be a bit clearer as, e.g., "name of translator used to process FITS header information"

mrawls commented on trail/upload/models.py:
```diff
@@ -3,29 +3,49 @@
+    standardizer_name = models.CharField("name of used translator.", max_length=20, null=False)
+    instrument = models.CharField("instrument name", max_length=20)
+    telescope = models.CharField("telescope", max_length=20, null=False)
+    science_program = models.CharField("science program image is a part of.", max_length=30)
```
Is there any way to indicate a field is required or not? I think "science program" is a nice to have, not an absolutely must have.

DinoBektesevic replied:
The `null` argument sets what is allowed to be null in the database. There is a more stringent requirement that sets what is required to be present at instantiation but I thought this was sufficient. 

In any case, the standardized dataclasses functionality has now been merged into the models so now we have `keys` and `required_keys` properties that allow inspecting what is required rather easily as well.

mrawls commented on trail/upload/models.py:
```diff
@@ -3,29 +3,49 @@
+
+class Metadata(models.Model):
+
+    # This will need to be fixed, cascading can orphan medata entries
```
Typo in comment (metadata entries, presumably). If this isn't a straightforward fix to do now, please open an issue.

DinoBektesevic replied:
This is me struggling against the Django ORM and schema layouts. Technically I think `CASCADING` would be fine here too, because it works the other way around, but I'm not sure how Django deals with orphaning in reverse so I had these protected. 

Of course, with how we are processing (in steps) we will need to address how we rollback changes and deal with cleanup. This however made thinking about the problem, for me, much harder so I operated under the assumption that all data we get will not fail just to get the code going. 

I think handling failures (both the very explicit ones like raised errors but also more subtle processing errors like badly processed data, or correct data but someone is trolling us with not-science images in data arrays) is really important but it's also confusing. This deserves an issue of its own where rollback and validation behavior is precisely defined. 

A part of the reason why this is so hard is that I'm not even sure what are all the failure cases. Now that we have something, this should definitely get tested out in anger and failures mapped out so that we know what our edge cases are. Maybe someone cleverer than me can reason about it in advance - in which case `FitsProcessor.process` method would be the most appropriate place to start implementing rollback behavior.

For most of the code in this PR I have assumed that no uploaded data will break the code in any non-trivial ways.

mrawls replied:
Hmmm... let's discuss more later, because I'd like to better understand the subtleties here. I think this is fine for now, just want to make sure we flag it to revisit when friendly data can't be assumed.

mrawls replied:
I swear I wrote a reply and GitHub ate it. Let's discuss more, because I'd like to understand the subtleties here better. For now, I'm OK with assuming uploaded data is friendly, but we'll need to break that assumption eventually.

DinoBektesevic replied:
Weird, I see both replies. Ok, issue https://github.com/dirac-institute/trailblazer/issues/15 - this will be added in a separate issue (as much as possible) and we will discuss details at a later date. I will mark discussions like these as resolved here.

mrawls commented on trail/upload/process_uploads/processors/rubin_processor.py:
```diff
@@ -0,0 +1,42 @@
+in their FITS files, but only a single CCD image, alongside a mask and a
+variance plane data.
+
+We don't want to process variance or mask planes.
```
... for now 😎

mrawls commented on trail/upload/process_uploads/standardized_dataclasses.py:
```diff
@@ -0,0 +1,208 @@
+                  ("required", bool),("closeEquality", bool)])
+
+wcsAttributes = np.array([
+    ("wcs_radius",   float, None, True, True),
```
The formatting here doesn't feel sufficiently PEP-8-ey to me, but I could be wrong. I saw one of your commits was "flakeify" so I'm assuming you ran some linter?

DinoBektesevic replied:
It's not, but I dare someone to codify and debug a layout in a non-column-aligned format. My preference here is to ignore it in flake. 

Re: general state of PEP 8 compliance - I _ran_ a linter and it printed stuff out about which I didn't do too much. 
Lintering should probably be sorted out as a part of a CI issue, before it gets out of hand.

mrawls replied:
I added a comment in #9 about this.

mrawls commented on trail/upload/process_uploads/standardized_dataclasses.py:
```diff
@@ -0,0 +1,208 @@
+        # asdict use here would be nice, but is much slower...
+        return {key: getattr(self, key) for key in self.keys}
+
+    def valid(self):
```
Maybe rename this to `isValid` to make it clearer this returns a boolean? You could get *real* bonus points if you used type checking (e.g., `def isValid(self) -> bool:`)

DinoBektesevic replied:
We can add typechecking everywhere, or nowhere. I'm not really for a mixed bag though. In any case, I dropped this method when I merged it with models in order to implement it at a later date when we sort out failure cases.

mrawls replied:
I'd kind of love to add it everywhere heheheh... it's so nice

DinoBektesevic replied:
Issue https://github.com/dirac-institute/trailblazer/issues/17

mrawls commented on trail/upload/tests/test_processors.py:
```diff
@@ -0,0 +1,166 @@
+
+class TemporaryUploadedFileWrapperTestCase(TestCase):
+    """Tests the TemporaryUploadedFileWrapper functionality. """
+    testDataDir = os.path.join(TESTDIR, "data")
```
Should we consider a separate repository for testdata so the main trailblazer project doesn't get too unwieldy to clone? Or has that ship sailed?

DinoBektesevic replied:
The image data has been cropped, the WCSs have been translated to fit the crop center and header keys adjusted appropriately. The total size of the dataset is 4MB, most of which in the DECam data (2MB), which could be much much smaller if I leave only a couple of cropped headers in it and make the rest a 1x1 array. If 4MB is too much we can move stuff out, otherwise I'm quite fine with it living here in this state. 

Perhaps if we get some kind of validation stuff going we will want actual full datasets at which point additional resources might be required.

mrawls commented on trail/upload/process_uploads/fits_processor.py:
```diff
@@ -0,0 +1,213 @@
+
+        # TODO: consider removing PIL dependency once trail detection is
+        # implemented, if it is implemented via OpenCV
+        normedImage = (normedImage.data*255).astype(np.uint8)
+        # this is grayscale
+        img = Image.fromarray(normedImage, "L")
```
Do we have to use PIL for this? Is it fast or otherwise super nice? If it would be straightforward, I'd love to use the color normalization I mentioned before, which you can find at https://github.com/lsst-dm/ap_pipe-notebooks/blob/master/plotLightcurve.py#L112-L144

DinoBektesevic replied:
AFAIK `matplotlib` can't be used to resize an image, so we need something else. Matplotlib will use PIL in the background anyhow, every time we [save an image ](https://matplotlib.org/stable/api/_as_gen/matplotlib.pyplot.imsave.html). 

PIL here is used just for resizing of the image, not its normalization, because it already comes with `matplotlib`. I guess we can transition to something like scikit image module for this but I wasn't sure if we want to install the entire scikit and scikit image extension for a single operation.
Eventually, if we implement some kind of image science data analysis stuff, I assume we will use a more powerful libraries like OpenCV or something. We can change this to cv resizing method then too. 

Resizing is not particularly slow, although there are magic keywords that can make it even faster. 

As far as image normalization goes: For all cases except DECam focal plane images, and even then only in cases when the file was created by the DECam Community Pipelines, the image normalization is performed by `Astropy.ImageNormalization` (see `FitsProcessor.normalizeImage`). 
We can change the normalization to use the asynh stretch, but that will only work well, for a particular choice of normalization constant, if we're processing calibrated exposures and will often not  be generally applicable, with the same normalization constant, for raw, just isr or fully calibrated image data. 

I picked histogram equalization because it's mathematically guaranteed to maximize contrast (which is what I thought I wanted when we are resizing down this much) but I'm not married to the idea. Again, see the comments in `FiitsProcessor.normalizeImage` method as it was left there as a potential `TODO` for anyone wanting to roll up their sleeves. 

As far as the DECam Community Pipeline image products - I'm not sure we can live with astropy's normalization there, see comments in `DecamFits.normalizeImage` processor method:

>      # astropy equalizer averages 4.3 seconds, the PIL approach can bring
>      # that down to cca 0.13. Without normalization the time is 0.04s

That's with the speedier histogram equalization, performance may take a larger hit.

mrawls replied:
OK great points, PIL it is! I just hadn't seen it before; I guess in Rubin we have all our own image manipulation C++ flavored shenanigans.

mrawls replied:
That's an excellent point about how the normalization I suggested may only work well for halfway calibrated images. We can leave it as histogram equalization to maximize contrast 👍

DinoBektesevic replied:
Yeah, there could be a particular normalization factor for asinh that works well across the board and we could add the normalization factor as an attribute, or a property for those processors that have multiple different appropriate normalization factors, and it would work just fine. It will make the images look nicer. 

Is this resolved, pending an issue creation or should we make this a requirement for this PR?

mrawls commented on trail/upload/process_uploads/fits_processor.py:
```diff
@@ -0,0 +1,213 @@
+    @classmethod
+    @abstractmethod
+    def canProcess(cls, uploadedFile, returnHdulist=False):
+        # docstring inherited from baseclass; TODO: check it's True
```
I'm a little lost in double-decorator base class land here. What docstring is this supposed to inherit?

DinoBektesevic replied:
Comment clarified, from `UploadProcessor`.

mrawls commented on trail/upload/process_uploads/fits_processor.py:
```diff
@@ -0,0 +1,213 @@
+        """
+        # All Header operations are done by Standardizers. HeaderStandardizer
+        # handles WCS standardization for all our data, so far.
+        raise NotImplemented()
```
I could be wrong, but I _think_ raising `NotImplementedError` is more appropriate than `NotImplemented` here.

DinoBektesevic replied:
Nice catch! Fixed everywhere where I made this blunder.

mrawls commented on trail/upload/process_uploads/standardizers/moa_standardizer.py:
```diff
@@ -0,0 +1,85 @@
+
+        return standardizedKeys
+
+    def standardizeWcs(self, **kwargs):
```
😱 do we really need to hard-code this? Two thoughts. Is there some kind of "None" or "nan" for postgres? Or, more at the heart of it, should we just reject the upload outright? Would it help to talk with the folks who sent me this image?

DinoBektesevic replied:
It would be great help to hear from people that sent the image to see how they kept track of where in the sky they were looking at because it wasn't clear to me at all. 

For the rest: see the comment about error handling [above](https://github.com/dirac-institute/trailblazer/pull/10#discussion_r654729848)

For cases exactly like this one we might even want to have a "commitable" or "processable" or an actual english word that means we accept or deny files like these - and then we won't have to wait to raise an error and skip wasting cycles on processing.

mrawls replied:
Emailed the source of the MOA image to ask about this just now, we'll see what she says.

mrawls commented on trail/upload/process_uploads/standardizers/moa_standardizer.py:
```diff
@@ -0,0 +1,85 @@
+        chip = self.header["CHIP"]
+        standardizedKeys["science_program"] = f"{run}-{field}-{filter}-{chip}"
+
+        # TODO: Fix datetimes
```
Oy, time is always hard. Please open an issue for this if you haven't.

DinoBektesevic replied:
Issue: https://github.com/dirac-institute/trailblazer/issues/16

mrawls commented on trail/upload/process_uploads/header_standardizer.py:
```diff
@@ -0,0 +1,329 @@
+
+    Standardization consists of:
+    * converting WCS information contained in each image-like header such as
+      PrimaryHDU, CompImageHDu etc. (i.e. each header for which it is infered
```
```suggestion
      PrimaryHDU, CompImageHDU etc. (i.e. each header for which it is infered
```

mrawls commented on trail/upload/process_uploads/header_standardizer.py:
```diff
@@ -0,0 +1,329 @@
+        additional metadata.
+    """
+
+    standardizers = dict()
+    """All registered header standardizers."""
+
+    name = None
+    """Standardizer's name. Only named standardizers will be registered."""
```
Please give me a bit more to go off of in doc-string land. Standardize, translate, and register are all words that mean very specific things around here, but this isn't making it super clear.

DinoBektesevic replied:
I've added some clarifying notes sections throughout doc strings for standardizers and processors but I'm open to suggestions.

In short every defined standardizer that sets a `name` will register itself with the top-level standardizer class as an standardizer that is available to standardize a header that meets its requirements in `canProcess`. This is what "registering itself" means. Maybe there's a better expression. The toplevel class dispatches the processing to the standardizer with the highest priority down the line. 

Same goes for processors.

mrawls commented on trail/upload/process_uploads/header_standardizer.py:
```diff
@@ -0,0 +1,329 @@
+        standardizedWcs = {}
+        centerX, centerY = int(dimX/2), int(dimY/2)
+
+        # TODO: When eventually logging is added to processing, redirect these
```
This one *definitely* needs an issue opened if we don't have one... I would argue we might initially want to give up if we can't make sense of the header WCS. Maybe later we can get smarter. Won't silently ignoring this lead to a metadata table full of useless WCS-less things?

DinoBektesevic replied:
See the comment about error handling [above](https://github.com/dirac-institute/trailblazer/pull/10#discussion_r654729848).

In short: absolutely yes, but it was not wholly clear to me what data is "good enough" to commit and what not.

mrawls commented on trail/upload/process_uploads/header_standardizer.py:
```diff
@@ -0,0 +1,329 @@
+
+        return standardizedWcs
+
+    # wow, do not flip these two decorators around...
```
👀

mrawls commented on trail/upload/process_uploads/header_standardizer.py:
```diff
@@ -0,0 +1,329 @@
+        Keyword arguments
+        -----------------
+        filename : `str`
+            Name of the file from which the HDU was read from, sometimes can encode
+            additional metadata.
```
Truer words were never spoken

mrawls commented on trail/upload/process_uploads/header_standardizer.py:
```diff
@@ -0,0 +1,329 @@
+        Raises
+        ------
+        ValueError
+            None of the registered processors can process the  upload.
```
```suggestion
            None of the registered processors can process the upload.
```

mrawls commented on trail/upload/process_uploads/header_standardizer.py:
```diff
@@ -0,0 +1,329 @@
+        for standardizer in cls.standardizers.values():
+            if standardizer.canStandardize(header):
+                standardizers.append(standardizer)
+        standardizers.sort(key=lambda standardizer: standardizer.priority, reverse=True)
```
Any way to rewrite this without a lambda? I just despise them is all.

DinoBektesevic replied:
Sure, it would look something like:

```
# ... for loop ...
def sort_by_priority(standardizer):
    return standardizer.priority
standardizers.sort(key=sort_by_priority, reverse=True)
```

if you're fine with that...

mrawls replied:
I can know that's what it's doing if I stare at it for 10 min. If it doesn't sacrifice speed, I prefer the clarity - thanks 😁

mrawls commented on trail/upload/process_uploads/header_standardizer.py:
```diff
@@ -0,0 +1,329 @@
+        standardizedKeys : `dict`
+          A dictionary with standardized header keys and values.
+        """
+        raise NotImplemented()
```
Clarify somewhere that it's not implemented because it's instrument-specific, please

DinoBektesevic replied:
I assume you want this for all abstract base classes methods? Because they will complain quite explicitly when someone tries to define a class without defining all the required abstract methods, I think something like: `Can't instantiate abstract class with abstract methods ...` thing.

mrawls replied:
Oh that's fair. I meant just in the docstring in these "standard" standardizers, saying explicitly they're ABC methods intended to be overridden (inherited? whatever the right word is) by an instrument-specific one instead. Maybe that's implied by how you wrote it and I'm just not as used to the convention as I should be.

DinoBektesevic replied:
No problems with adding the comments, I just was not clear if you meant this one specific method or all abstract methods in the code?

mrawls replied:
Ideally all of them. Because I'm bad at telling which ones are this flavor vs an actual functionality we haven't implemented yet 🙂 It can be the same generic one line comment.

DinoBektesevic replied:
Ok, will add tonight. None of the methods that currently raise `NotImplementedError` are raising them because it's a thing that isn't actually implemented yet. For some, like MOA, they are fake and should be fixed, for others they actually work. Mostly they raise them because they are meant to be overridden once inherited from the class, at which point they inherit the docstring of the method too. Or so I hope (the docstring bit).

mrawls commented on trail/upload/process_uploads/header_standardizer.py:
```diff
@@ -0,0 +1,329 @@
+
+        Parameters
+        ----------
+        hdu : `obhect` or `None`, optional
```
```suggestion
        hdu : `object` or `None`, optional
```

mrawls commented on trail/upload/process_uploads/header_standardizer.py:
```diff
@@ -0,0 +1,329 @@
+
+        Parameters
+        ----------
+        hdu : `obhect` or `None`, optional
```
```suggestion
        hdu : `object` or `None`, optional
```

mrawls commented on trail/upload/process_uploads/header_standardizer.py:
```diff
@@ -0,0 +1,329 @@
+        cornerDec = cornerSkyCoord.dec.to(u.deg)
+
+        unitSphereCenter = np.array([
+            np.cos(centerDec) * np.cos(centerRa),
```
You've forced all the Ras and Decs to be in degrees, and then done trigonometry. Are you 100% sure that's right?

DinoBektesevic replied:
Nice catch, thank god for Astropy though because fixing this otherwise would require fixing the test data values too which would be a pain.

```
from astropy import units as u
from astropy.coordinates import SkyCoord
c = SkyCoord(ra=10.625*u.degree, dec=41.2*u.degree, frame='icrs')

c.ra
# <Longitude 10.625 deg>
c.ra.to(u.deg)
# <Longitude 10.625 deg>

import numpy as np
np.cos(c.ra)
# <Quantity 0.98285499>
np.cos(c.ra.to(u.deg))
# <Quantity 0.98285499>
np.cos(c.ra.to(u.rad))
# <Quantity 0.98285499>
```

mrawls commented on trail/upload/process_uploads/processors/decam_processor.py:
```diff
@@ -0,0 +1,360 @@
+        largePath = os.path.join(self.media_root, self.uploadedFile.basename+'_plane_large.jpg')
+
+        smallThumb = self._createFocalPlaneImage(smallPlane)
+        plt.imsave(smallPath, smallThumb.planeImage.T, pil_kwargs={"quality":30})
+        # due to the size of these images immediately release memory
+        del smallThumb
+
+        largeThumb = self._createFocalPlaneImage(largePlane)
+        plt.imsave(largePath, largeThumb.planeImage.T, pil_kwargs={"quality":30})
```
Policy for file/image overwriting error-handling situations?

DinoBektesevic replied:
See the comment about error handling [above](https://github.com/dirac-institute/trailblazer/pull/10#discussion_r654729848).

Is overwriting an error or not? I'm not sure in what cases it should be allowed, or should updating data be allowed only by manual cranking it. 
Is processing a file that produces the same standardized metadata an error - our standardized metadata is basically the observer location, observation time and observing instrument so this is not really sufficient to tell (f.e. DECam via Rubin will even have the same observing times but we can't tell they're the same pointing, different ccd at the moment) so we need to check if WCS matches I guess. Then I guess if WCS and metadata matches do we want to soft delete, hard delete or substitute in-place? Anyhow... 

The newest code centralizes the place were we save the images, in preparation for the move to S3. I assume we can make a unique string (read as: randomly generated string large enough to give an asymptotic guarantee of never duplicating) and centralizing that functionality for all processors too, as a quick fix, but there's a good chance this behavior will become clear once we answer above questions so I'm not particularly keen on resolving it in this PR. 

For this PR I've assumed no uploaded file ever will break any of the code in it in non-trivial ways.

mrawls replied:
Yeah I'm not sure either 😄 we need a duplicates policy in general. If all the metadata (including WCS) matches, I think we'd want to default to not accepting the upload so we don't waste our upload/download cloud $ on duplicates. Does that sound reasonable?

DinoBektesevic replied:
This sounds perfectly reasonable to me in this case. I just didn't have the capacity to think about all of the details here. To boot it felt like too much work that will leave more people pending on the merge of this PR.
Punted to issue https://github.com/dirac-institute/trailblazer/issues/15

mrawls commented on trail/upload/process_uploads/processors/multi_extension_fits.py:
```diff
@@ -0,0 +1,86 @@
+                    isinstance(hdu, ImageHDU))):
+            return False
+
+        # People store all kind of stuff even in ImageHDUs, let's make sure we
+        # don't crash the server by saving 120k x 8000k table disguised as an
+        # image (I'm looking at you SDSS!)
```
Good thinking... yikes

mrawls commented on trail/upload/process_uploads/processors/single_extension_fits.py:
```diff
@@ -0,0 +1,51 @@
+
+    def storeThumbnails(self):
+        large, small = self._createThumbnails(self.uploadedFile.basename, self.imageData)
+        plt.imsave(small["savepath"], small["thumb"], pil_kwargs={"quality": 10})
+        plt.imsave(large["savepath"], large["thumb"], pil_kwargs={"quality": 30})
```
Again, question about what to do if this would be overwriting an existing image/file

DinoBektesevic replied:
See the comment about error handling [above](https://github.com/dirac-institute/trailblazer/pull/10#discussion_r654734302).

mrawls commented on trail/upload/process_uploads/standardized_dataclasses.py:
```diff
@@ -0,0 +1,208 @@
+    _requiredKeys: ClassVar[list] = metaAttrs[metaAttrs["required"] == True]["names"]
+    _optionalKeys: ClassVar[list] = metaAttrs[metaAttrs["required"] == False]["names"]
+
+    def isCloseTo(self, other, **kwargs):
```
Docstring plz, in fact the whole Mixin class could use some more docstrings

DinoBektesevic replied:
This was merged with models and file trashed, models now have docstrings.

mrawls reviewed (commented):
That was a marathon of a review! Impressive work, let me know if any of my comments are unclear. I trust you to merge when you think you've addressed them.

DinoBektesevic committed:
Merge standardized dataclasses functionality into models.

DinoBektesevic committed:
Merge standardized dataclasses with models. Add thumbnails.

Merge standardized dataclasses functionality with database models.

Add thumbnails class in preparation for gallery upgrade to
querying the database for images.
Make processed thumbnail images grayscale instead of viridis.

Minor refactoring of processors and standardizers. Got rid of
unused methods, transitioned from returning dictionaries to
returning model classes.

Fix tests.

DinoBektesevic committed:
Rebase on top of main

DinoBektesevic committed:
Add tests for models and dataclass.

DinoBektesevic committed:
Fix last of the review comments.

DinoBektesevic commented:
Ok, fixed the last comments and added tests for Models. 

The remainder of issues here were punted other issues in which we will tackle them. Mainly all is well except: we don't have good error handling, data update/replace/remove logic and MOA-II WCS is a placeholder pending error handling, or getting Astrometry.net solver working.

Looking for any last comments?

mrawls commented:
Understood. Let's get it merged so we can find fun new ways to break things 😄
